minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,823 @@
+                continue
+
+            yield ZNode(
+                coord=ccoord,
+                shape=self._shape,
+                convert_to_z=convert,
+            )
+
+    def odd_neighbors_generator(
+        self,
+        where: Callable[[CartesianCoord | ZephyrCoord], bool] = lambda coord: True,
+    ) -> Generator[ZNode]:
+        """Generator of odd neighbors of self when restricted by ``where``.
+        Args:
+            where (Callable[[CartesianCoord | ZephyrCoord], bool]):
+                A coordinate filter. Applies to ``ccoord`` if :py:attr:`self.convert_to_z` is ``False``,
+                or to ``zcoord`` if :py:attr:`self.convert_to_z` is ``True``. Defaults to always.
+        Yields:
+            ZNode: Odd neighbors of self when restricted by ``where``.
+        """
+        x, y, k = self._ccoord
+        convert = self.convert_to_z
+        changing_index = 1 if x % 2 == 0 else 0
+        for s in [-2, 2]:
+            new_x = x + s if changing_index == 0 else x
+            new_y = y + s if changing_index == 1 else y
+            ccoord = CartesianCoord(x=new_x, y=new_y, k=k)
+
+            # Check ccoord is valid. If not valid, ignore this ccoord
+            try:
+                self._check_ccoord_val(ccoord)
+                self._check_ccoord_shape(ccoord)
+            except ValueError:
+                continue
+            
+            coord = ccoord if not convert else cartesian_to_zephyr(ccoord)
+            if not where(coord):
+                continue
+
+            yield ZNode(
+                coord=ccoord,
+                shape=self._shape,
+                convert_to_z=convert,
+            )
+
+    def neighbors_generator(
+        self,
+        nbr_kind: EdgeKind | Iterable[EdgeKind] | None = None,
+        where: Callable[[CartesianCoord | ZephyrCoord], bool] = lambda coord: True,
+    ) -> Generator[ZNode]:
+        """Generator of neighbors of self when restricted by ``nbr_kind`` and ``where``.
+
+        Args:
+            nbr_kind (EdgeKind | Iterable[EdgeKind] | None, optional):
+                Edge kind filter. Restricts yielded neighbors to those connected by the given edge kind(s).
+                If ``None``, no filtering is applied. Defaults to None.
+            where (Callable[[CartesianCoord | ZephyrCoord], bool]):
+                A coordinate filter. Applies to ``ccoord`` if :py:attr:`self.convert_to_z` is ``False``,
+                or to ``zcoord`` if :py:attr:`self.convert_to_z` is ``True``. Defaults to always.
+
+        Yields:
+            ZNode: Neighbors of self when restricted by ``nbr_kind`` and ``where``.
+        """
+        if nbr_kind is None:
+            kinds = {kind for kind in EdgeKind}
+        elif isinstance(nbr_kind, EdgeKind):
+            kinds = {nbr_kind}
+        else:
+            kinds = set(nbr_kind)
+        if EdgeKind.INTERNAL in kinds:
+            for cc in self.internal_neighbors_generator(where=where):
+                yield cc
```

Comment:
Perfect use case for `yield from` (here, and below):

```suggestion
            yield from self.internal_neighbors_generator(where=where)
```